internal/evaluation_target/application_snapshot_image/application_snapshot_image.go

```diff
@@ -0,0 +1,234 @@
+// Copyright 2022 Red Hat, Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+//
+// SPDX-License-Identifier: Apache-2.0
+
+package application_snapshot_image
+
+import (
+	"context"
+	"encoding/json"
+	"errors"
+	"fmt"
+	"os"
+	"path"
+	"strings"
+
+	"github.com/google/go-containerregistry/pkg/name"
+	ecc "github.com/hacbs-contract/enterprise-contract-controller/api/v1alpha1"
+	"github.com/in-toto/in-toto-golang/in_toto"
+	"github.com/sigstore/cosign/cmd/cosign/cli/rekor"
+	"github.com/sigstore/cosign/pkg/cosign"
+	"github.com/sigstore/cosign/pkg/oci"
+	"github.com/sigstore/cosign/pkg/policy"
+	"github.com/sigstore/cosign/pkg/signature"
+	cosignTypes "github.com/sigstore/cosign/pkg/types"
+	"k8s.io/apimachinery/pkg/types"
+
+	"github.com/hacbs-contract/ec-cli/internal/evaluator"
+	"github.com/hacbs-contract/ec-cli/internal/kubernetes_client"
+	"github.com/hacbs-contract/ec-cli/internal/policy_source"
+)
+
+// ConftestNamespace is the default rego namespace using for policy evaluation
+const ConftestNamespace = "release.main"
+
+// PipelineRunBuildType is the type of attestation we're interested in evaluating
+const PipelineRunBuildType = "https://tekton.dev/attestations/chains/pipelinerun@v2"
+
+var newConftestEvaluator = evaluator.NewConftestEvaluator
+var kubernetesCreator = kubernetes_client.NewKubernetes
+
+// ApplicationSnapshotImage represents the structure needed to evaluate an Application Snapshot Image
+type ApplicationSnapshotImage struct {
+	reference    name.Reference
+	checkOpts    cosign.CheckOpts
+	attestations []oci.Signature
+	Evaluator    *evaluator.ConftestEvaluator
+}
+
+// NewApplicationSnapshotImage returns an ApplicationSnapshotImage struct with reference, checkOpts, and evaluator ready to use.
+func NewApplicationSnapshotImage(ctx context.Context, image string, publicKey string, rekorURL string, policyConfiguration string) (*ApplicationSnapshotImage, error) {
```

Comment:
This is _really_ awkward, we had a nice split between image signature and attestation verification and policy evaluation, now it's lumped together. This now leads to multiple redundant invocations, e.g. `FetchEnterpriseContractPolicy` is now invoked for each image rather than once. There is a lot of duplicate code here between this and `internal/image/validate.go`. The code as it should look here is closer in `internal/image/validate.go`.

Reply:
Fetching the policy for each image sounds like a bug to fix. Can we create a Jira to track that?

Reply:
I've moved the policy fetching out of this function so it's only invoked once.